cmd/kubeadm/app/util/etcd/etcd.go

```diff
@@ -83,40 +86,127 @@ FlagLoop:
 	return true, nil
 }
 
-// GetStatus gets server status
-func (c GenericClient) GetStatus() (*clientv3.StatusResponse, error) {
-	const dialTimeout = 5 * time.Second
+// New creates a new EtcdCluster client
+func New(endpoints []string, ca, cert, key string) (*Client, error) {
+	client := Client{Endpoints: endpoints}
+
+	if ca != "" || cert != "" || key != "" {
+		tlsInfo := transport.TLSInfo{
+			CertFile:      cert,
+			KeyFile:       key,
+			TrustedCAFile: ca,
+		}
+		tlsConfig, err := tlsInfo.ClientConfig()
+		if err != nil {
+			return nil, err
+		}
+		client.TLS = tlsConfig
+	}
+
+	return &client, nil
+}
+
+// NewFromStaticPod creates a GenericClient from the given endpoints, manifestDir, and certificatesDir
+func NewFromStaticPod(endpoints []string, manifestDir string, certificatesDir string) (*Client, error) {
+	hasTLS, err := PodManifestsHaveTLS(manifestDir)
+	if err != nil {
+		return nil, fmt.Errorf("could not read manifests from: %s, error: %v", manifestDir, err)
+	}
+	if hasTLS {
+		return New(
+			endpoints,
+			filepath.Join(certificatesDir, constants.EtcdCACertName),
+			filepath.Join(certificatesDir, constants.EtcdHealthcheckClientCertName),
+			filepath.Join(certificatesDir, constants.EtcdHealthcheckClientKeyName),
+		)
+	}
+	return New(endpoints, "", "", "")
+}
+
+// GetVersion returns the etcd version of the cluster.
+// An error is returned if the version of all endpoints do not match
+func (c Client) GetVersion() (string, error) {
+	var clusterVersion string
+
+	versions, err := c.GetClusterVersions()
+	if err != nil {
+		return "", err
+	}
+	for _, v := range versions {
+		if clusterVersion == "" {
+			// This is the first version we've seen
+			clusterVersion = v
+		} else if v != clusterVersion {
+			return "", fmt.Errorf("etcd cluster contains endpoints with mismatched versions: %v", versions)
+		} else {
+			clusterVersion = v
+		}
```

Comment:
change only if see fit :+1: 
```
for _, v := range versions {
	if clusterVersion != "" && clusterVersion != v {
		return "", fmt.Errorf("etcd cluster contains endpoints with mismatched versions: %v", versions)
	}
	clusterVersion = v
}
```

Reply:
I added this to my PR for fixing the static pod upgrade regression in master.